### PR TITLE
feat: add slot-name-casing rule

### LINKED ISSUE #2580: Rule proposal: `slot-name-casing`
**Please describe what the rule should do:**
The proposed linter rule should allow limiting the casing for slot names.
For example in bigger projects it might happen that people mix camelCase and kebab-case for the names of component slots. To keep the codebase uniform it would be really helpful if a linter rule would warn or optionally even automatically fix invalid slot names according to the projects configuration.

**What category should the rule belong to?**

[x] Enforces code style (layout)
[ ] Warns about a potential error (problem)
[ ] Suggests an alternate way of doing something (suggestion)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule should warn about:**

Bad:
```vue
<template>
  <div>
    <slot name="random-Slot_name" />
  </div>
</template
```

Good:
```vue
<template>
  <div>
    <slot name="randomSlotName" />
  </div>
</template
```

Good:
```vue
<template>
  <div>
    <slot name="random-slot-name" />
  </div>
</template
```

**Additional context**
It would most likely be enough to detect invalid slot names when a slot is defined. While fixing invalid usage of slots would be nice for existing projects adapting the rule but might also blow up it's complexity.

### PR DESCRIPTION
**Pull request by waynzh**
resolve #2580

### CODE REVIEW

**Review by ota-meshi — APPROVED** [review]
Looks good to me! Thank you!

File: lib/rules/slot-name-casing.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+/** @type {OptionType[]} */
+const allowedCaseOptions = ['camelCase', 'kebab-case', 'singleword']
+
+module.exports = {
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'enforce specific casing for the slot name',
```
Comment (FloEdelmann) [inline lib/rules/slot-name-casing.js:31]:
```suggestion
      description: 'enforce specific casing for slot names',
```

(please run `npm run update` afterwards)

File: docs/rules/slot-name-casing.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+# vue/slot-name-casing
+
+> enforce specific casing for the slot name
+
+- :exclamation: <badge text="This rule has not been released yet." vertical="middle" type="error"> _**This rule has not been released yet.**_ </badge>
+
+## :book: Rule Details
+
+This rule enforce proper casing of slot names in vue components.
```
Comment (FloEdelmann) [inline docs/rules/slot-name-casing.md:16]:
```suggestion
This rule enforces proper casing of slot names in Vue components.
```

File: tests/lib/rules/slot-name-casing.js [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    `<template><slot name="fooBar" /></template>`,
+    `<template><slot :name="fooBar" /></template>`,
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+      <slot name="foo" />
+      <slot name="foo-bar" />
+      <slot :name="fooBar" />
+      </template>`,
```
Comment (FloEdelmann) [inline tests/lib/rules/slot-name-casing.js:32]:
```suggestion
        <template>
          <slot name="foo" />
          <slot name="foo-bar" />
          <slot :name="fooBar" />
        </template>
      `,
```

File: tests/lib/rules/slot-name-casing.js [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+      options: ['kebab-case']
+    },
+    {
+      filename: 'test.vue',
+      code: `
+      <template>
+      <slot name="foo" />
+      <slot :name="fooBar" />
+      </template>`,
+      options: ['singleword']
```
Comment (FloEdelmann) [inline tests/lib/rules/slot-name-casing.js:42]:
```suggestion
      code: `
        <template>
          <slot name="foo" />
          <slot :name="fooBar" />
        </template>
      `,
      options: ['singleword']
```

File: tests/lib/rules/slot-name-casing.js [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        },
+        {
+          messageId: 'invalidCase',
+          data: {
+            name: 'foo-Bar_baz',
+            caseType: 'camelCase'
+          },
+          line: 4,
+          column: 13
+        }
```
Comment (FloEdelmann) [inline tests/lib/rules/slot-name-casing.js:71]:
```suggestion
      code: `
        <template>
          <slot name="foo-bar" />
          <slot name="foo-Bar_baz" />
        </template>
      `,
      errors: [
        {
          messageId: 'invalidCase',
          data: {
            name: 'foo-bar',
            caseType: 'camelCase'
          },
          line: 3,
          column: 17
        },
        {
          messageId: 'invalidCase',
          data: {
            name: 'foo-Bar_baz',
            caseType: 'camelCase'
          },
          line: 4,
          column: 17
        }
```

File: tests/lib/rules/slot-name-casing.js [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        },
+        {
+          messageId: 'invalidCase',
+          data: {
+            name: 'foo-Bar_baz',
+            caseType: 'kebab-case'
+          },
+          line: 4,
+          column: 13
+        }
```
Comment (FloEdelmann) [inline tests/lib/rules/slot-name-casing.js:100]:
```suggestion
      code: `
        <template>
          <slot name="fooBar" />
          <slot name="foo-Bar_baz" />
        </template>
      `,
      options: ['kebab-case'],
      errors: [
        {
          messageId: 'invalidCase',
          data: {
            name: 'fooBar',
            caseType: 'kebab-case'
          },
          line: 3,
          column: 17
        },
        {
          messageId: 'invalidCase',
          data: {
            name: 'foo-Bar_baz',
            caseType: 'kebab-case'
          },
          line: 4,
          column: 17
        }
```

File: tests/lib/rules/slot-name-casing.js [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        },
+        {
+          messageId: 'invalidCase',
+          data: {
+            name: 'foo-Bar_baz',
+            caseType: 'singleword'
+          },
+          line: 5,
+          column: 13
+        }
```
Comment (FloEdelmann) [inline tests/lib/rules/slot-name-casing.js:139]:
```suggestion
      code: `
        <template>
          <slot name="foo-bar" />
          <slot name="fooBar" />
          <slot name="foo-Bar_baz" />
        </template>
      `,
      options: ['singleword'],
      errors: [
        {
          messageId: 'invalidCase',
          data: {
            name: 'foo-bar',
            caseType: 'singleword'
          },
          line: 3,
          column: 17
        },
        {
          messageId: 'invalidCase',
          data: {
            name: 'fooBar',
            caseType: 'singleword'
          },
          line: 4,
          column: 17
        },
        {
          messageId: 'invalidCase',
          data: {
            name: 'foo-Bar_baz',
            caseType: 'singleword'
          },
          line: 5,
          column: 17
        }
```

**Review by FloEdelmann — CHANGES_REQUESTED** [review]
Generally looks good to me, but I have a few minor suggestions.

File: lib/rules/slot-name-casing.js (hunk 1 above)
Comment (waynzh) [reply lib/rules/slot-name-casing.js]: Done ✅

**Review by FloEdelmann — APPROVED** [review]
Thank you :slightly_smiling_face:

**Pull request merged**